lib/repl.js

```diff
@@ -296,15 +322,60 @@ function REPLServer(prompt,
           return;
         }
       }
-    }
 
-    // After executing the current expression, store the values of RegExp
-    // predefined properties back in `savedRegExMatches`
-    for (var idx = 1; idx < savedRegExMatches.length; idx += 1) {
-      savedRegExMatches[idx] = RegExp[`$${idx}`];
+      if (awaitPromise && !err) {
+        let sigintListener;
+        pause();
+        const promise = self.breakEvalOnSigint ? Promise.race([
+          result,
+          new Promise((resolve, reject) => {
+            sigintListener = () => {
+              reject(new Error('Script execution interrupted.'));
+            };
+            priorizedSigintQueue.push(sigintListener);
+          })
+        ]) : result;
+
+        promise.then((result) => {
+          if (self.breakEvalOnSigint) {
+            // Remove priorized SIGINT listener if it was not called.
+            const idx = priorizedSigintQueue.indexOf(sigintListener);
+            if (idx >= 0) {
+              priorizedSigintQueue.splice(idx, 1);
+            }
+          }
+
+          finishExecution(undefined, result);
+          unpause();
+        }, (err) => {
+          if (self.breakEvalOnSigint) {
+            // Remove priorized SIGINT listener if it was not called.
+            const idx = priorizedSigintQueue.indexOf(sigintListener);
+            if (idx >= 0) {
+              priorizedSigintQueue.splice(idx, 1);
+            }
+          }
```

Comment:
If you turned `priorizedSigintQueue` into a Set(), you could replace these blocks with a one-liner:
```js
priorizedSigintQueue.delete(sigintListener);
```
Or is it important that the queue maintains ordering?

Reply:
I don't *think* the queue ordering is important, but since Set is ordered anyway I reckon that would work.

Reply:
Done.